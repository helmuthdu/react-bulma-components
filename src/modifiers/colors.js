/* eslint-disable no-unused-vars */
import cn from 'classnames';
import PropTypes from 'prop-types';
import CONSTANTS from '../constants';

const colors = [...Object.values(CONSTANTS.COLORS), ...Object.values(CONSTANTS.GREY_COLORS)];

export default {
  propTypes: {
    backgroundColor: PropTypes.oneOf(colors),
    textColor: PropTypes.oneOf(colors)
  },
  defaultProps: {},
  getClassName: props =>
    cn({
      [`has-background-${props.backgroundColor}`]: props.backgroundColor,
      [`has-text-${props.textColor}`]: props.textColor
    }),
  clean: ({ textColor, backgroundColor, ...props }) => props
};
