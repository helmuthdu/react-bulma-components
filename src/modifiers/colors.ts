import cn from 'classnames';
import { Colors, Greyscale } from '../constants';

export type ColorsModifier = {
  backgroundColor?: Colors & Greyscale;
  textColor?: Colors & Greyscale;
};

export default {
  defaultProps: {},
  getClassName: (props: any) =>
    cn({
      [`has-background-${props.backgroundColor}`]: props.backgroundColor,
      [`has-text-${props.textColor}`]: props.textColor
    }),
  clean: ({ textColor, backgroundColor, ...props }: ColorsModifier) => props
};
