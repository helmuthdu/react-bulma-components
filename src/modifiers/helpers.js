/* eslint-disable no-unused-vars */
import cn from 'classnames';
import PropTypes from 'prop-types';

export default {
  propTypes: {
    clearfix: PropTypes.bool,
    clipped: PropTypes.bool,
    hidden: PropTypes.bool,
    invisible: PropTypes.bool,
    marginless: PropTypes.bool,
    overlay: PropTypes.bool,
    paddingless: PropTypes.bool,
    pull: PropTypes.oneOf(['right', 'left']),
    radiusless: PropTypes.bool,
    shadowless: PropTypes.bool,
    unselectable: PropTypes.bool
  },
  defaultProps: {
    clearfix: false,
    clipped: false,
    hidden: false,
    invisible: false,
    marginless: false,
    overlay: false,
    paddingless: false,
    radiusless: false,
    shadowless: false,
    unselectable: false
  },
  getClassName: props =>
    cn({
      'is-clearfix': props.clearfix,
      'is-clipped': props.clipped,
      'is-hidden': props.hidden,
      'is-invisible': props.invisible,
      'is-marginless': props.marginless,
      'is-overlay': props.overlay,
      'is-paddingless': props.paddingless,
      'is-radiusless': props.radiusless,
      'is-shadowless': props.shadowless,
      'is-unselectable': props.unselectable,
      [`is-pulled-${props.pull}`]: props.pull
    }),
  clean: ({
    clearfix,
    clipped,
    hidden,
    invisible,
    marginless,
    overlay,
    paddingless,
    pull,
    radiusless,
    shadowless,
    unselectable,
    ...props
  }) => props
};
