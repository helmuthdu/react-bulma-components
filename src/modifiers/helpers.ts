import clsx from 'clsx';
import { Display } from '../constants';

export type HelpersModifier = {
  clearfix?: boolean;
  clipped?: boolean;
  display?: Display;
  hidden?: boolean;
  invisible?: boolean;
  overlay?: boolean;
  pulled?: 'right' | 'left';
  radiusless?: boolean;
  relative?: boolean;
  scrolless?: boolean;
  shadowless?: boolean;
  unselectable?: boolean;
};

export default {
  defaultProps: {
    clearfix: false,
    clipped: false,
    hidden: false,
    invisible: false,
    overlay: false,
    radiusless: false,
    relative: false,
    scrolless: false,
    shadowless: false,
    unselectable: false
  },
  getClassName: (props: HelpersModifier) =>
    clsx({
      'is-clearfix': props.clearfix,
      'is-clipped': props.clipped,
      'is-hidden': props.hidden,
      'is-invisible': props.invisible,
      'is-overlay': props.overlay,
      'is-radiusless': props.radiusless,
      'is-relative': props.relative,
      'is-scrolless': props.scrolless,
      'is-shadowless': props.shadowless,
      'is-unselectable': props.unselectable,
      [`is-${props.display}`]: props.display,
      [`is-pulled-${props.pulled}`]: props.pulled
    }),
  clean: ({
    clearfix,
    clipped,
    display,
    hidden,
    invisible,
    overlay,
    pulled,
    radiusless,
    relative,
    scrolless,
    shadowless,
    unselectable,
    ...props
  }: HelpersModifier) => props
};
