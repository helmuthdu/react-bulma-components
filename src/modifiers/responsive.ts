/* eslint-disable no-unused-vars */
import clsx from 'clsx';
import { Breakpoints, Display, Gap, Spacing, TextAlignment } from '../constants';

type BreakpointSize = {
  display?: Display | { only: Display };
  gap?: Gap;
  hidden?: boolean | { only: boolean };
  textSize?: Spacing | { only: Spacing };
  textAlignment?: TextAlignment;
};

export type BreakpointsSize = { [key in Breakpoints]?: BreakpointSize };

export type ResponsiveModifier = BreakpointsSize;

const getValue = (val: any) => val?.only ?? val;
const getOnly = (val: any) => (val?.only ? '-only' : '');

const getSizeClassFromProp = (sizes: BreakpointsSize = {}) =>
  Object.entries(sizes)
    .filter(([key, values]) => values)
    .reduce(
      // @ts-ignore
      (classes, [breakpoint, { display, gap, hidden, textAlignment, textSize }]) => ({
        ...classes,
        'is-variable': gap !== undefined,
        [`is-${gap}-${breakpoint}`]: gap !== undefined,
        [`has-text-${getValue(textAlignment)}-${breakpoint}${getOnly(textAlignment)}`]: textAlignment,
        [`is-${getValue(display)}-${breakpoint}${getOnly(display)}`]: display,
        [`is-hidden-${breakpoint}${getOnly(hidden)}`]: hidden,
        [`is-size-${textSize}-${breakpoint}`]: textSize !== undefined
      }),
      {}
    );

export default {
  defaultProps: {},
  getClassName: ({ mobile, tablet, desktop, widescreen, fullhd, touch }: ResponsiveModifier) =>
    clsx({
      ...getSizeClassFromProp({ mobile, tablet, desktop, widescreen, fullhd, touch })
    }),
  clean: ({ mobile, tablet, desktop, widescreen, fullhd, touch, ...props }: ResponsiveModifier) => props
};
