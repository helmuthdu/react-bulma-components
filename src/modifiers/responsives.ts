/* eslint-disable no-unused-vars */
import clsx from 'clsx';

type SizeShape = {
  display?: {
    value?: 'block' | 'flex' | 'inline' | 'inline-block' | 'inline-flex';
    only?: boolean;
  };
  hide?: {
    value?: boolean;
    only?: boolean;
  };
  textSize?: {
    value?: 1 | 2 | 3 | 4 | 5 | 6;
  };
  textAlignment?: {
    value?: 'centered' | 'justified' | 'left' | 'right';
    only?: boolean;
  };
};

export type ResponsiveModifier = {
  responsive?: {
    mobile?: SizeShape;
    tablet?: SizeShape;
    desktop?: SizeShape;
    widescreen?: SizeShape;
    fullhd?: SizeShape;
    touch?: SizeShape;
  };
};

const getSizeClassFromProp = (sizes: { [key: string]: SizeShape }) =>
  Object.keys(sizes).reduce((classes, size) => {
    const display = sizes[size].display || {};
    const hide = sizes[size].hide || {};
    const textAlignment = sizes[size].textAlignment || {};
    const textSize = sizes[size].textSize || {};

    const obj = {
      ...classes,
      [`has-text-${textAlignment.value}-${size}${textAlignment.only ? '-only' : ''}`]: textAlignment.value,
      [`is-${display.value}-${size}${display.only ? '-only' : ''}`]: display.value,
      [`is-hidden-${size}${hide.only ? '-only' : ''}`]: hide.value,
      [`is-size-${textSize.value}-${size}`]: (textSize.value || -1) > 0
    };

    return obj;
  }, {});

export default {
  defaultProps: {},
  getClassName: (props: any) =>
    clsx({
      ...getSizeClassFromProp(props.responsive || {})
    }),
  clean: ({ responsive, ...props }: ResponsiveModifier) => props
};
