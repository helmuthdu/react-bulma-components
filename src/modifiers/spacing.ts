import clsx from 'clsx';
import { Spacing } from '../constants';

export type SpacingModifier = {
  padding?: Spacing;
  paddingTop?: Spacing;
  paddingLeft?: Spacing;
  paddingBottom?: Spacing;
  paddingRight?: Spacing;
  paddingHorizontal?: Spacing;
  paddingVertical?: Spacing;
  margin?: Spacing;
  marginTop?: Spacing;
  marginLeft?: Spacing;
  marginBottom?: Spacing;
  marginRight?: Spacing;
  marginHorizontal?: Spacing;
  marginVertical?: Spacing;
};

const modifier = {
  defaultProps: {},
  getClassName: ({
    margin,
    marginBottom,
    marginHorizontal,
    marginLeft,
    marginRight,
    marginTop,
    marginVertical,
    padding,
    paddingBottom,
    paddingHorizontal,
    paddingLeft,
    paddingRight,
    paddingTop,
    paddingVertical
  }: SpacingModifier): string =>
    clsx({
      [`p-${padding}`]: padding,
      [`pt-${paddingTop}`]: paddingTop,
      [`pl-${paddingLeft}`]: paddingLeft,
      [`pb-${paddingBottom}`]: paddingBottom,
      [`pr-${paddingRight}`]: paddingRight,
      [`px-${paddingHorizontal}`]: paddingHorizontal,
      [`py-${paddingVertical}`]: paddingVertical,
      [`m-${margin}`]: margin,
      [`mt-${marginTop}`]: marginTop,
      [`ml-${marginLeft}`]: marginLeft,
      [`mb-${marginBottom}`]: marginBottom,
      [`mr-${marginRight}`]: marginRight,
      [`mx-${marginHorizontal}`]: marginHorizontal,
      [`my-${marginVertical}`]: marginVertical
    }),
  clean: ({
    padding,
    paddingTop,
    paddingLeft,
    paddingBottom,
    paddingRight,
    paddingHorizontal,
    paddingVertical,
    margin,
    marginTop,
    marginLeft,
    marginBottom,
    marginRight,
    marginHorizontal,
    marginVertical,
    ...props
  }: SpacingModifier) => props
};

export default modifier;
