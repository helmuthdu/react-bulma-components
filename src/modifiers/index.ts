import clsx from 'clsx';
import * as React from 'react';
import colors, { ColorsModifier } from './colors';
import helpers, { HelpersModifier } from './helpers';
import responsive, { ResponsiveModifier } from './responsives';
import spacing, { SpacingModifier } from './spacing';
import typography, { TypographyModifier } from './typography';

const compose = (...fns: any[]) => (args: any) => fns.reduce((arg: any, fn) => fn(arg), args);

export type ElementModifier = ColorsModifier &
  HelpersModifier &
  ResponsiveModifier &
  SpacingModifier &
  TypographyModifier & {
    className?: any;
    children?: React.ReactNode;
    as?: keyof JSX.IntrinsicElements | React.ComponentType<any>;
    style?: React.CSSProperties;
    testId?: string | string[];
  };

export default {
  defaultProps: {
    ...colors.defaultProps,
    ...helpers.defaultProps,
    ...responsive.defaultProps,
    ...spacing.defaultProps,
    ...typography.defaultProps
  },
  getClassName: (props: any) =>
    clsx(
      colors.getClassName(props),
      helpers.getClassName(props),
      responsive.getClassName(props),
      spacing.getClassName(props),
      typography.getClassName(props)
    ),
  clean: (props: any) => compose(colors.clean, helpers.clean, responsive.clean, spacing.clean, typography.clean)(props)
};
