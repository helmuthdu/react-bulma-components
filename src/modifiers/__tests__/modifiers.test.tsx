import modifiers from '..';

describe('Helper proptypes', () => {
  it('should have clearfix classes', () => {
    expect(
      modifiers.getClassName({
        clearfix: true
      })
    ).toMatchSnapshot();
  });

  it('should have text helpers', () => {
    expect(
      modifiers.getClassName({
        textColor: 'success',
        textAlignment: 'centered',
        italic: true,
        textTransform: 'uppercase',
        textWeight: 'bold'
      })
    ).toMatchSnapshot();
  });

  it('should have responsive modifier flex-tablet-only and block-widescreen', () => {
    expect(
      modifiers.getClassName({
        tablet: {
          display: {
            only: 'flex'
          }
        },
        widescreen: {
          display: 'block'
        }
      })
    ).toMatchSnapshot();
  });

  it('should have hidden modifier tablet-only and widescreen', () => {
    expect(
      modifiers.getClassName({
        tablet: {
          hidden: { only: true }
        },
        widescreen: {
          hidden: true
        }
      })
    ).toMatchSnapshot();
  });

  it('should have alignment responsive modifiers', () => {
    expect(
      modifiers.getClassName({
        tablet: {
          textAlignment: 'centered'
        },
        widescreen: {
          textAlignment: 'left'
        }
      })
    ).toMatchSnapshot();
  });

  it('should have text color success', () => {
    expect(
      modifiers.getClassName({
        textColor: 'success'
      })
    ).toMatchSnapshot();
  });

  it('should have background color success', () => {
    expect(
      modifiers.getClassName({
        backgroundColor: 'success'
      })
    ).toMatchSnapshot();
  });
});
