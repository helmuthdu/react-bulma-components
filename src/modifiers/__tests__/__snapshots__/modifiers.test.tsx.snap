// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Helper proptypes should have alignment responsive modifiers 1`] = `""`;

exports[`Helper proptypes should have background color success 1`] = `"has-background-success"`;

exports[`Helper proptypes should have clearfix classes 1`] = `"is-clearfix"`;

exports[`Helper proptypes should have hidden modifier tablet-only and widescreen 1`] = `""`;

exports[`Helper proptypes should have responsive modifier flex-tablet-only and block-widescreen 1`] = `""`;

exports[`Helper proptypes should have text color success 1`] = `"has-text-success"`;

exports[`Helper proptypes should have text helpers 1`] = `"has-text-success has-text-centered has-text-weight-bold is-uppercase is-italic"`;
